integration-tests/singleton_flush_test.go

```diff
@@ -0,0 +1,237 @@
+package main
+
+import (
+	"encoding/xml"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"gopkg.in/yaml.v2"
+)
+
+type SingletonFlushTestSuite struct {
+	suite.Suite
+	ServiceID    string
+	ServiceToken string
+	AppIDs       []string
+	PlanID       string
+	Metrics      []Metric
+}
+
+func (suite *SingletonFlushTestSuite) SetupSuite() {
+	suite.ServiceID = "test-service-id"
+	suite.AppIDs = []string{"test-app-id-1", "test-app-id-2"}
+	suite.ServiceToken = "test-service-token"
+	suite.PlanID = "test-plan-id"
+	suite.Metrics = []Metric{
+		{"hits", "1", []UsageLimit{
+			{Day, 10000},
+		}},
+		{"rq", "2", []UsageLimit{
+			{Month, 10000},
+		}},
+	}
+
+}
+
+func (suite *SingletonFlushTestSuite) BeforeTest(suiteName, testName string) {
+	if testName == "TestSingletonContainerFlush" {
+		configErr := configureSingletonFlush("ContainerLimit")
+		require.Nilf(suite.T(), configErr, "Error configuring envoy.yaml for container flush : %v", configErr)
+
+		require.Eventually(suite.T(), func() bool {
+			fmt.Printf("Creating service with service_id: %s, service_token:%s", suite.ServiceID, suite.ServiceToken)
+			serviceErr := CreateService(suite.ServiceID, suite.ServiceToken)
+			if serviceErr != nil {
+				return false
+			}
+			return true
+		}, 4*time.Second, 500*time.Millisecond, "Error creating the service")
+
+		for _, app := range suite.AppIDs {
+			require.Eventually(suite.T(), func() bool {
+				err := AddApplication(suite.ServiceID, app, suite.PlanID)
+				if err != nil {
+					return false
+				}
+				return true
+			}, 4*time.Second, 500*time.Millisecond, "Error creating the app")
+		}
+
+		require.Eventually(suite.T(), func() bool {
+			err := AddMetrics(suite.ServiceID, &suite.Metrics)
+			if err != nil {
+				return false
+			}
+			return true
+		}, 4*time.Second, 500*time.Millisecond, "Error adding metrics")
+
+		require.Eventually(suite.T(), func() bool {
+			err := UpdateUsageLimits(suite.ServiceID, suite.PlanID, &suite.Metrics)
+			if err != nil {
+				return false
+			}
+			return true
+		}, 4*time.Second, 500*time.Millisecond, "Error adding usage limits")
+	}
+
+}
+
+func configureSingletonFlush(flushMode string) error {
+	yamlFile, fileErr := ioutil.ReadFile("envoy.yaml")
+	if fileErr != nil {
+		return fileErr
+	}
+	yamlData := make(map[interface{}]interface{})
+	yamlErr := yaml.Unmarshal(yamlFile, &yamlData)
+	if yamlErr != nil {
+		return yamlErr
+	}
+	var singletonConfig = fmt.Sprintf(`
+      name: envoy.bootstrap.wasm
+      typed_config:
+        '@type': type.googleapis.com/envoy.extensions.wasm.v3.WasmService
+        singleton: true
+        config:
+          name: "singleton_service"
+          root_id: "singleton_service"
+          configuration: 
+            "@type": type.googleapis.com/google.protobuf.StringValue
```

Comment:
So much code just to change 1 value. 🤔

Reply:
Yeah this is how you usually do yaml formatting in golang. With the golang limitation of unmarshalling, you can either define a struct with the related tags or dynamically handle with interfaces. For very large yaml, since it's difficult to define a struct, dynamic handling is done. But for deep nested, to prevent the nested reads with interfaces, whole section is formatted.